substrate/frame/contracts/src/benchmarking/call_builder.rs

```diff
@@ -0,0 +1,104 @@
+use crate::{
+	benchmarking::{Contract, WasmModule},
+	exec::Stack,
+	storage::meter::Meter,
+	BalanceOf, Config, Determinism, GasMeter, Origin, Schedule, TypeInfo, WasmBlob, Weight,
+};
+use codec::{Encode, HasCompact};
+use core::fmt::Debug;
+use sp_core::Get;
+use sp_std::prelude::*;
+
+type StackExt<'a, T> = Stack<'a, T, WasmBlob<T>>;
+
+/// A LeakReclaimer is a guard that reclaims the leaked box when dropped.
+struct LeakReclaimer<T> {
+	ptr: *mut T,
+}
+
+impl<T> LeakReclaimer<T> {
+	/// Creates a new `LeakReclaimer`, leaking a boxed value and returning a static mutable
+	/// reference to its content, along with the guard for later cleanup.
+	///
+	/// # Safety
+	///
+	/// The caller must ensure that the returned reference is not used after the `LeakReclaimer` is
+	/// dropped.
+	unsafe fn new(value: T) -> (&'static mut T, Self) {
+		let mut boxed = Box::new(value);
+		let ptr: *mut T = &mut *boxed;
+		let static_ref: &'static mut T = Box::leak(boxed);
+
+		(static_ref, Self { ptr })
+	}
+}
+
+impl<T> Drop for LeakReclaimer<T> {
+	fn drop(&mut self) {
+		// Safety: The pointer is valid as it points to the leaked box, created by `[Self::new]`.
+		unsafe {
+			let _ = Box::from_raw(self.ptr);
+		}
+	}
+}
+
+/// A builder use to prepare a benchmark contracts call.
+pub struct CallBuilder<T: Config> {
+	caller: T::AccountId,
+	dest: T::AccountId,
+	gas_meter: GasMeter<T>,
+	storage_meter: Meter<T>,
+	schedule: Schedule<T>,
+	input: Vec<u8>,
+}
+
+impl<T> CallBuilder<T>
+where
+	T: Config + pallet_balances::Config,
+	<BalanceOf<T> as HasCompact>::Type: Clone + Eq + PartialEq + Debug + TypeInfo + Encode,
+{
+	/// Create a new builder for the given module.
+	pub fn new(module: WasmModule<T>) -> Self {
+		let instance = Contract::<T>::new(module.clone(), vec![]).unwrap();
+		let caller = instance.caller.clone();
+		let dest = instance.account_id.clone();
+
+		Self {
+			caller,
+			dest,
+			schedule: T::Schedule::get(),
+			gas_meter: GasMeter::new(Weight::MAX),
+			storage_meter: Default::default(),
+			input: vec![],
+		}
+	}
+
+	/// Prepare a call to the module.
+	/// Returns a a closure used to invoke the call.
+	pub fn build(self) -> impl FnOnce() {
```

Comment:
Hack: 
I need to leak and reclaim some references here as the benchmark needs to move ownership of the code behind `#[block]` not sure if there is a better approach

Reply:
I think you wouldn't need to leak anything or use `unsafe` Rust code if you change the API of your `CallBuilder`. Instead of a `fn build() -> impl FnOnce` method you could provide a `fn run()` method that simply executes the test instead of returning a closure to allow the caller to execute the test. So the `CallBuilder` itself can simply store all the data. I did something similar with Wasmi's translation tests: https://github.com/wasmi-labs/wasmi/blob/master/crates/wasmi/src/engine/translator/tests/op/block.rs#L98 and https://github.com/wasmi-labs/wasmi/blob/master/crates/wasmi/src/engine/translator/tests/driver.rs.

Reply:
in the code you shared, it looks like you are also creating the module inside the `run` fn, which we want to do ahead of time here so that it's done outside of the benchmarked block.

Reply:
On further note: In my opinion returning `impl X` usually also is a code smell in my opinion and should only ever be used for highly generic code (which again is kinda code smelly) if there is clearly no other possible way out.

Reply:
> , I already see leaks, for example if the returned closure is accidentally never executed for whatever reason and simply dropped which again is a problem

right, I guess we could just mark this as unsafe to put the caller in charge of upholding these safety guarantees

> On further note: In my opinion returning impl X usually also is a code smell in my opinion and should only ever be used for some generic code

I had it setup a bit differently on the previous [commit](https://github.com/paritytech/polkadot-sdk/pull/3915/commits/d381d65992c7d0913fc65456479f6731a52c4af1), but ideally, because this code will be repeated a lot in the benchmark we can come up with an API as concise as the current one.

Reply:
@ggwpez, if you have any suggestions for resolving the lifetime issues we're encountering, I'd be eager to hear your thoughts

Reply:
To give more context, the issue here is that the benchmark framework generate this function

```rust
fn instance(
    &self,
    components: &[(frame_benchmarking::BenchmarkParameter, u32)],
    verify: bool,
) -> Result<frame_benchmarking::__private::Box<dyn FnOnce() -> Result<(), frame_benchmarking::BenchmarkError>>, frame_benchmarking::BenchmarkError> 
{
    let r = components
        .iter()
        .find(|&c| c.0 == frame_benchmarking::BenchmarkParameter::r)
        .ok_or("Could not find component during benchmark preparation.")?
        .1;

    let invoke = CallBuilder::<T>::new(WasmModule::getter("seal0", "seal_now", r)).build();
    Ok(
        frame_benchmarking::__private::Box::new(move || -> Result<(), frame_benchmarking::BenchmarkError> {
            {
                invoke();
            }
            if verify {}
            Ok(())
        }),
    )
}
```

Without tricking Rust with 'static ref we end up with the classic
```
returns a value referencing data owned by the current function
```

Reply:
I guess an approach could be to update the impl_benchmark macro here 
https://github.com/paritytech/polkadot-sdk/blob/2d466329c7212311a43b6dcc51df7708db4f0f14/substrate/frame/benchmarking/src/v1.rs?plain=1#L1111-L1117

instead of returning a closure, we would just "inject" the start & end value around the benchmark block.